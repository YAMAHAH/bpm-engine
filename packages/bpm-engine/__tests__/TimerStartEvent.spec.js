import fs from 'fs';

import BPMEngine from 'bpm-engine';
import History from './Plugins/History';
import sleep from './Plugins/sleep';

describe('TimerStartEvent', () => {
  it('Can start a process with a timer', async () => {
    const history = new History();
    const engine = new BPMEngine({
      plugins: [history],
    });

    await engine.deployWorkflowDefinition({
      xml: fs.readFileSync(`${__dirname}/diagrams/TimerStartEvent.bpmn`, 'utf-8'),
    });

    await sleep(4000);

    // we expect 3 process instances to have been started while waiting for 4 seconds
    // because the timerStart event interval is: R3/${timestamp}/PT1S which equals to
    // repeat 3 times, starting on timestamp of deployment, every 1 second
    expect(engine.persist.store.processInstances.length).toBe(3);
    expect(history).toMatchSnapshot();
  });
});
