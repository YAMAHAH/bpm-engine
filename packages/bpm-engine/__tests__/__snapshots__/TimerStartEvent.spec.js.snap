// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerStartEvent Can start a process with a timer 1`] = `
History {
  "onActive": [Function],
  "onComplete": [Function],
  "onReady": [Function],
  "store": Array [
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "StartEvent_1",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "Task_1me2wuc",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "ready",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "active",
    },
    Object {
      "elementId": "EndEvent_073fq1n",
      "lanes": undefined,
      "participant": undefined,
      "state": "complete",
    },
  ],
}
`;
